### PR TITLE
added Adahessian optimizer

### PR DESCRIPTION
**Pull request by amirgholami**
This pull request adds AdaHessian optimizer: 

https://arxiv.org/abs/2006.00719

### CODE REVIEW

**Comment by jettify** [comment]
thanks a lot! I will take a look soon.

File: torch_optimizer/adahessian.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+                           '\t\t\t  loss.backward(), make sure the option create_graph is\n' +
+                           '\t\t\t  set to True.')
+
+        v = [2 * torch.randint_like(p, high=2) - 1 for p in params]
+
+        # this is for distributed setting with single node and multi-gpus, 
+        # for multi nodes setting, we have not support it yet.
+        if not self.single_gpu:
+            for v1 in v:
+                dist.all_reduce(v1)
```
Comment (jettify) [inline torch_optimizer/adahessian.py:83]: I think this variable is used before declaration, I guess better to remove multi gpu support, since it is broken anyway.

**Comment by jettify** [comment]
I am running pytorch test suite and several tests are failing:
```python
tests/test_optimizer.py::TestOptim::test_optimizer[Adahessian_0] FAILED
tests/test_optimizer.py::TestOptim::test_optimizer[Adahessian_1] FAILED
tests/test_optimizer.py::TestOptim::test_optimizer[Adahessian_2] PASSED
tests/test_optimizer.py::TestOptim::test_optimizer[Adahessian_3] PASSED
tests/test_optimizer.py::TestOptim::test_optimizer[Adahessian_4] FAILED
tests/test_optimizer.py::TestOptim::test_optimizer[Adahessian_5] FAILED
tests/test_optimizer.py::TestOptim::test_optimizer[Adahessian_6] FAILED
```
with errors like:
```
E           assert False
E            +  where False = <built-in method allclose of type object at 0x10f408c30>(tensor([[ 1.3770, -0.6990, -1.1279,  0.8318,  1.2768],\n        [-0.4016, -1.7553, -0.7445, -1.8927, -0.8778],\n        ....0364,  1.2505, -1.2547,  0.2096, -1.0199],\n        [-0.4855,  0.4929,  0.6671,  0.2404, -1.2928]], requires_grad=True), tensor([[ 1.3770, -0.6990, -1.1279,  0.8319,  1.2768],\n        [-0.4018, -1.7554, -0.7446, -1.8928, -0.8779],\n        ....0364,  1.2505, -1.2547,  0.2096, -1.0199],\n        [-0.4855,  0.4929,  0.6671,  0.2404, -1.2928]], requires_grad=True), atol=0.0001)
E            +    where <built-in method allclose of type object at 0x10f408c30> = torch.allclose
```

**Comment by jettify** [comment]
Broken multi gpu, type annotation and bunch of lint errors easy to fix and not a blocker. But we need to figure out source of why weights are different in pytorch test suite before merging. You can repro by registering optimizer here:
https://github.com/jettify/pytorch-optimizer/blob/233c3a6b6643cfb07541524072f7335ee840b831/tests/test_optimizer.py#L67-L95

**Comment by amirgholami** [comment]
> Broken multi gpu, type annotation and bunch of lint errors easy to fix and not a blocker. But we need to figure out source of why weights are different in pytorch test suite before merging. You can repro by registering optimizer here:
> 
> https://github.com/jettify/pytorch-optimizer/blob/233c3a6b6643cfb07541524072f7335ee840b831/tests/test_optimizer.py#L67-L95

Could you please let me know what command line you used to run test_optimizer.py?

Thanks,
-Amir

**Comment by jettify** [comment]
Just regular purest:
```
py.test -sv tests/test_optimizer.py -k Adahessian
```

**Comment by jettify** [comment]
```
174  	        # Load state dict
175  	        state_dict = deepcopy(optimizer.state_dict())
176  	        state_dict_c = deepcopy(optimizer.state_dict())
177  	        optimizer_c.load_state_dict(state_dict_c)
178
179  	        precision = 0.0001
180  	        # Run both optimizations in parallel
181  	        for _i in range(20):
182  	            optimizer.step(fn)
183  	            optimizer_c.step(fn_c)
184  ->	            assert torch.allclose(weight, weight_c, atol=precision)
185  	            assert torch.allclose(bias, bias_c, atol=precision)
```
test basically clones state of the optimizer into new one, and new optimizer with same state is expected to produce similar results as old one.

**Comment by jettify** [comment]
Just in case here is my branch, based on this one that unblocks CI https://github.com/jettify/pytorch-optimizer/pull/247 you can see relevant test failure here https://github.com/jettify/pytorch-optimizer/pull/247/checks?check_run_id=1592485269

**Comment by amirgholami** [comment]
> Just in case here is my branch, based on this one that unblocks CI #247 ([45c1f99](https://github.com/jettify/pytorch-optimizer/commit/45c1f99ba9bb686ea59dec0c1c7e2412ea7251c7))
> 
> as result you can see relevant test failure on CI here https://github.com/jettify/pytorch-optimizer/pull/247/checks?check_run_id=1592485269

Thanks for the quick follow up. I think I know why the test fails. AdaHessian uses randomized linear algebra to estimate Hessian trace. To do this every time you call the optimizer, it draws a random vector v and applies it to the Hessian:

https://github.com/jettify/pytorch-optimizer/blob/amirgholami-adahessian/torch_optimizer/adahessian.py#L96

When I set this random vector to a fixed value then I do not get that error anymore (although I get an error about weights not being on the same device). 

P.S: We do support Multi-GPU training. The comment in the code only applies to cases where you have multiple clusters of GPUs. However, AdaHessian works fine when you run it on multiple GPUs on the same machine (say an 8 GPU system)

File: torch_optimizer/adahessian.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        """
+
+        # Check backward was called with create_graph set to True
+        for i, grad in enumerate(grads):
+            if grad.grad_fn is None:
+                raise RuntimeError('Gradient tensor {:} does not have grad_fn. When calling\n'.format(i) +
+                           '\t\t\t  loss.backward(), make sure the option create_graph is\n' +
+                           '\t\t\t  set to True.')
+
+        v = [2 * torch.randint_like(p, high=2) - 1 for p in params]
```
Comment (jettify) [inline torch_optimizer/adahessian.py:77]: I think it is good ideas to seed this random generator for reproducible results. For example add seed parameter to class constructor and use it here.

**Comment by jettify** [comment]
Good point I missed that `random_like` function, I think it should be possible to do following:

1) seed random number generator for reproducible results.
2) use `torch.random.get_rng_state` / `torch.random.set_rng_state` to satisfy pytroch optimizer tests, as result cloned optimizer should produce same result as test expects.

File: torch_optimizer/adahessian.py (hunk 1 above)
Comment (jettify) [reply torch_optimizer/adahessian.py:83]: Do not see where `dist` is declared, this is why I think multi gpu possible not working

File: torch_optimizer/adahessian.py (hunk 2 above)
Comment (jettify) [reply torch_optimizer/adahessian.py:77]: Also possibly `memory_format=torch.preserve_format`

**Comment by amirgholami** [comment]
@jettify I have added the random seed. Could you please have a look?

**Comment by jettify** [comment]
Can you please also rebase with recent master? to address conflict in README

**Comment by amirgholami** [comment]
> Can you please also rebase with recent master? to address conflict in README

I actually did that in [9cc73f6](https://github.com/jettify/pytorch-optimizer/pull/239/commits/9cc73f67a7b8b26b38f55b1898d360593dc4bffb) and the only difference now is the addition of adahessian information in the README file.

**Comment by jettify** [comment]
Was able to resave conflict via GitHub UI.

**Comment by jettify** [comment]
I will take a look over holidays, want to see if I manage to do `torch.random.get_rng_state/torch.random.set_rng_state` thing and get that pytorch case passing. If not gonna merge current approach.

**Comment by amirgholami** [comment]
> I will take a look over holidays, want to see if I manage to do `torch.random.get_rng_state/torch.random.set_rng_state` thing and get that pytorch case passing. If not gonna merge current approach.

Ok thanks. When I checked the values in the weights and biases were the same between the two optimizers.

P.S: I just fixed an issue the rng. The seed needs to be set in the Adahessian constructor and not inside GetTrace function, as there we need iid random numbers.

**Comment by jettify** [comment]
`torch.manual_seed(seed)` is setting seed globally, so this may interfere if someone doing the same somewhere else in the code. 
Here is few other points:
1) I think `seed=None` is better default, so if user of optimiser does not specify it we use system one.
2) in tests for repro purposes we want set `seed=0`
3) Ideally we can instance RNG and use for optimizer purposes without changing system one. But I am not 100% sure how this works with cuda, but definitely gonna work for cpu.

**Comment by amirgholami** [comment]
> `torch.manual_seed(seed)` is setting seed globally, so this may interfere if someone doing the same somewhere else in the code.
> Here is few other points:
> 
> 1. I think `seed=None` is better default, so if user of optimiser does not specify it we use system one.
> 2. in tests for repro purposes we want set `seed=0`
> 3. Ideally we can instance RNG and use for optimizer purposes without changing system one. But I am not 100% sure how this works with cuda, but definitely gonna work for cpu.

That is a fair point. I just fixed 1 and 2, and pushed the code.

**Comment by jettify** [comment]
Tried to make pytorch test case without luck. Test case clones state of optimizer and runs first one and clone in same time, expected similar results after 25 iterations. I managed to make cloned RNG ( it is not picklable by default), but still optimizers diverge after 12 iteration or so.

**Comment by jettify** [comment]
Anyway decided to merge, since other tests looks good. Thank you for the Adahessian!
